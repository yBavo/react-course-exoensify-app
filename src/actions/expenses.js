import uuid from 'uuid';
import database from '../firebase/firebase'

/**
 * action types
 */
export const ADD_EXPENSE = "ADD_EXPENSE"
export const REMOVE_EXPENSE = "REMOVE_EXPENSE"
export const EDIT_EXPENSE = "EDIT_EXPENSE"

/**
 * action creators
 */
// Add an expense
export const addExpense = (expense) => ({
	type: ADD_EXPENSE,
	expense
})

export const startAddExpense = (expenseData = {}) => {
	return (dispatch) => {
		const {
			description = '',
			note = '',
			amount = 0,
			createdAt = 0
		} = expenseData;
		const expense = { description, note, amount, createdAt }

		return database.ref('expenses').push(expense).then((ref) => {
			dispatch(addExpense({
				id: ref.key,
				...expense
			}))
		})
	}
}

// Remove an expense
export const removeExpense = ({ id } = {}) => ({
	type: REMOVE_EXPENSE,
	id
})
// Start remove an expense
export const startRemoveExpense = ({ id }) => {
	return (dispatch) => {
		return database.ref(`expenses/${id}`)
			.remove()
			.then(() => {
				dispatch(removeExpense({id}))
			})
	}
}

// Edit an expense
export const editExpense = (id, updates) => ({
	type: EDIT_EXPENSE,
	id,
	updates
})
// Start edit expense
export const startEditExpense = (id, updates) => {
	return (dispatch) => {
		return database.ref(`expenses/${id}`)
		.update(updates)
		.then(() => {
			dispatch(editExpense(id, updates))
		})
	}
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
	type: 'SET_EXPENSES',
	expenses
});

export const startSetExpenses = () => {
	return (dispatch) => {
		return database.ref('expenses')
			.once('value')
			.then((snapshot) => {
				const expenses = [];

				snapshot.forEach((childSnapshot) => {
					expenses.push({
						id: childSnapshot.key,
						...childSnapshot.val()
					})
				})

				dispatch(setExpenses(expenses))
			})
			.catch((e) => {
				console.log('Error!', e)
			})
	}
}