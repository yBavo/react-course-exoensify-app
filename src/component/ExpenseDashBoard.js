import React from 'react';
import ExpenseListFilter from './ExpenseListFilters'
import ExpenseList from './ExpenseList'
import ExpensesSummary from './ExpensesSummary'


const ExpenseDashBoardPage = () => (
	<div>
		<h1>Expense DashBoard</h1>
		<div>
		<ExpensesSummary />
		<ExpenseListFilter />
		<ExpenseList />
		</div>
	</div>
);

export default ExpenseDashBoardPage;
