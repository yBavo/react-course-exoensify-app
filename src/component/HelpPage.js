import React from 'react';

const HelpPage = () => (
	<div>
		<h1>Help Page</h1>
	</div>
);

export default HelpPage;
