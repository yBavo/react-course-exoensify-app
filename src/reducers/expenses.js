import {
	ADD_EXPENSE,
	REMOVE_EXPENSE,
	EDIT_EXPENSE
} from '../actions/expenses'

const initialState = []

export default (state = initialState, action) => {
	switch (action.type) {
		case ADD_EXPENSE:
			return [
				...state,
				action.expense
			];
		case REMOVE_EXPENSE:
			return state.filter(({ id }) => action.id !== id);
		case EDIT_EXPENSE:
			return state.map((expense) => {
				if (expense.id === action.id) {
					return {
						...expense,
						...action.updates
					};
				} else {
					return expense;
				}
			});
		default:
			return state;
	}
}