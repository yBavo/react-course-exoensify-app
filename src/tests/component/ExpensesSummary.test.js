import React from 'react'
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { ExpensesSummary } from '../../component/ExpensesSummary';
import expenses from '../fixtures/expenses';

configure({ adapter: new Adapter() });

test('Should render ExpensesSummary with  1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);
  expect(wrapper).toMatchSnapshot();
})

test('Should render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={235258} />);
  expect(wrapper).toMatchSnapshot();
})