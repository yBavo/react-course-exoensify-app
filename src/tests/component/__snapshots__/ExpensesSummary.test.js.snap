// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpensesSummary with  1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    2.35$
  </h1>
</div>
`;

exports[`Should render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    2,352.58$
  </h1>
</div>
`;
